### PR TITLE
Catalog2Reg for China

### PR DESCRIPTION
**Pull request by cristinamanum**
Changes to allow registration blobs to be added to China.

There are two changes in this PR:
1. The registration blobs will point to flat container instead of packages. (for this change FlatContainerPackagePathProvider was introduced)
2. The catalog2Reg waits for the catalog2dnx (for this now the job takes an extra parameter for the flatcontainer cursor)

### CODE REVIEW

File: src/Catalog/CatalogItem.cs [hunk 1]
```diff
@@ -10,11 +10,25 @@ namespace NuGet.Services.Metadata.Catalog
 {
     public abstract class CatalogItem
     {
+        Uri _baseAddress;
```
Comment (skofman1) [inline src/Catalog/CatalogItem.cs]: private

File: src/Catalog/CatalogWriterBase.cs [hunk 2]
```diff
@@ -148,14 +154,16 @@ protected virtual ResourceSaveOperation CreateSaveOperationForItem(IStorage stor
...
+            var catalogItemUri = item.GetItemAddress();
 
             var operation = new ResourceSaveOperation();
-            operation.ResourceUri = resourceUri;
+            operation.ResourceUri = catalogItemUri;
 
             if (content != null)
             {
-                operation.SaveTask = storage.Save(resourceUri, content, cancellationToken);
+                //the save operation will not be done through the customDomain but directly through the storage
```
Comment (skofman1) [inline src/Catalog/CatalogWriterBase.cs]: nit: capital letter in the beginning of comment, like in the rest of the method

File: src/Catalog/CatalogItem.cs [hunk 3]
```diff
@@ -37,5 +51,23 @@ public virtual IGraph CreateContentGraph(CatalogContext context)
         {
             return null;
         }
+
+        /// <summary>
+        /// Changes the host of the baseAddress with the customDomain
+        /// </summary>
+        /// <param name="baseAddress"></param>
```
Comment (skofman1) [inline src/Catalog/CatalogItem.cs]: nit: remove empty documentation tags

File: src/Catalog/CatalogItem.cs [hunk 4]
```diff
@@ -37,5 +51,23 @@ public virtual IGraph CreateContentGraph(CatalogContext context)
...
+        /// <param name="customDomain"></param>
+        /// <returns></returns>
+        public static  Uri AliasesAddress(Uri baseAddress, string customDomain)
```
Comment (skofman1) [inline src/Catalog/CatalogItem.cs]: I would suggest a clearer name: RelaceHostForUri or something..

File: src/Catalog/CatalogItem.cs [hunk 5]
```diff
@@ -10,11 +10,25 @@ namespace NuGet.Services.Metadata.Catalog
...
         public DateTime TimeStamp { get; set; }
 
         public Guid CommitId { get; set; }
 
-        public Uri BaseAddress { get; set; }
+        public Uri BaseAddress
+        {
+            get
+            {
+                return AliasesAddress(_baseAddress, CustomDomain);
```
Comment (skofman1) [inline src/Catalog/CatalogItem.cs]: every time BaseAddress is called AliasesAddress will be recalculated. To improve perf you can calculate this ones, on _baseAddress set

File: src/Catalog/Registration/RegistrationMaker.cs [hunk 6]
```diff
@@ -20,7 +20,8 @@ public static class RegistrationMaker
             Uri contentBaseAddress,
             int partitionSize,
             int packageCountThreshold,
-            CancellationToken cancellationToken)
+            CancellationToken cancellationToken,
+            string customDomain)
```
Comment (skofman1) [inline src/Catalog/Registration/RegistrationMaker.cs]: nit: the convention in the code is that CancellationToken is the last parameter

File: src/Catalog/Registration/RegistrationMakerCatalogWriter.cs [hunk 7]
```diff
@@ -86,18 +91,21 @@ protected override ResourceSaveOperation CreateSaveOperationForItem(IStorage sto
             if (registrationMakerCatalogItem != null)
             {
                 var content = item.CreateContent(Context); // note: always do this first
-                var resourceUri = item.GetItemAddress();
+                var itemAddress = item.GetItemAddress();
+                //the storage will indicate the location where the content will be saved
```
Comment (skofman1) [inline src/Catalog/Registration/RegistrationMakerCatalogWriter.cs]: nit: same comment for capital letter for comments

File: src/Catalog/Registration/RegistrationPersistence.cs [hunk 8]
```diff
@@ -23,15 +23,23 @@ public class RegistrationPersistence : IRegistrationPersistence
         private readonly RecordingStorage _storage;
         private readonly Uri _registrationBaseAddress;
         private readonly Uri _contentBaseAddress;
-
-        public RegistrationPersistence(StorageFactory storageFactory, RegistrationKey registrationKey, int partitionSize, int packageCountThreshold, Uri contentBaseAddress)
+        private readonly string _customDomain ;
```
Comment (skofman1) [inline src/Catalog/Registration/RegistrationPersistence.cs]: nit: space

File: src/Catalog/CatalogWriterBase.cs [hunk 9]
```diff
@@ -18,8 +18,13 @@ public abstract class CatalogWriterBase : IDisposable
     {
         protected List<CatalogItem> _batch;
         protected bool _open;
+        protected string _customDomain;
 
-        public CatalogWriterBase(IStorage storage, ICatalogGraphPersistence graphPersistence = null, CatalogContext context = null)
+        public CatalogWriterBase(
```
Comment (skofman1) [inline src/Catalog/CatalogWriterBase.cs]: do those changes impact Feed2Catalog job?

File: src/Catalog/CatalogItem.cs (hunk 4 above)
Comment (skofman1) [inline src/Catalog/CatalogItem.cs]: I didn't see any UTs for this method.

File: src/Catalog/CatalogWriterBase.cs (hunk 2 above)
Comment (scottbommarito) [inline src/Catalog/CatalogWriterBase.cs]: also space between the text of the comment and the `//`

for example: `// The save...`

File: src/Catalog/CatalogWriterBase.cs (hunk 9 above)
Comment (scottbommarito) [inline src/Catalog/CatalogWriterBase.cs]: I was confused at first as well, but `RegistrationMakerCatalogWriter` extends this class.

File: src/Catalog/Registration/FlatContainerPackagePathProvider.cs [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace NuGet.Services.Metadata.Catalog.Registration
+{
+    public class FlatContainerPackagePathProvider : IPackagePathProvider
+    {
+        public string GetPackagePath(string id, string version)
+        {
+            version = NuGetVersionUtility.NormalizeVersion(version);
+
+            return $"v3-flatcontainer/{id.ToLowerInvariant()}/{version.ToLowerInvariant()}/{id.ToLowerInvariant()}.{version.ToLowerInvariant()}.nupkg";
```
Comment (scottbommarito) [inline src/Catalog/Registration/FlatContainerPackagePathProvider.cs]: This should use the V3 index.json to determine the base address of the flat-container. There's not guarantee that we will always use `v3-flatcontainer` or that we will only have a single flat-container.

File: src/Ng/Jobs/Catalog2RegistrationJob.cs [hunk 11]
```diff
@@ -78,12 +80,24 @@ protected override void Init(IDictionary<string, string> arguments, Cancellation
                 source,
                 storageFactories.LegacyStorageFactory);
 
-            RegistrationMakerCatalogItem.PackagePathProvider = new PackagesFolderPackagePathProvider();
+            if (isContentFlatContainer)
+            {
+                var flatContainerCursorUriString = arguments.GetOrThrow<string>(Arguments.FlatContainerCursorUri);
```
Comment (scottbommarito) [inline src/Ng/Jobs/Catalog2RegistrationJob.cs]: You can very easily calculate the flat-container's cursor uri by using the flat-container's base address. You shouldn't need to pass it in via a command line argument.

File: src/Catalog/CatalogWriterBase.cs [hunk 12]
```diff
@@ -183,7 +191,7 @@ protected virtual StorageContent CreateIndexContent(IGraph graph, Uri type)
         {
             IGraph graph = new Graph();
 
-            INode resourceNode = graph.CreateUriNode(resourceUri);
+            INode resourceNode = graph.CreateUriNode(CatalogItem.AliasesAddress(resourceUri, _customDomain));
```
Comment (scottbommarito) [inline src/Catalog/CatalogWriterBase.cs]: These changes seem very scary to me.

Basically, in the past, all had were resource uris with the same host, so we passed these uris around on their own. Now, however, we care about the host, and so we have added tons of places where we are replacing the host.

If we missed a place in which the host should have been replaced, we could very easily get into a scenario where some uri somewhere is being written with an incorrect host. Additionally, in the future, if we want to provide some additional modifications to the uri, we could potentially need to modify each of these places again, and could again miss.

I think we should rethink this design. I would much rather see something in which we stop passing around uris and start passing around a `ResourceIdentity` object that we can ask for the uri of which would perform the `AliasesAddress` transform. By changing interfaces and methods that previously used uris to `ResourceIdentity`s, we can guarantee through strong typing that the uris are being transformed when expected.

I understand doing that would probably be a much bigger change, but I think it would be a lot more stable in the long run.

File: src/Ng/Jobs/LightningJob.cs [hunk 13]
```diff
@@ -401,7 +442,7 @@ private void InitStrike(IDictionary<string, string> arguments)
                             if (graph.Nodes.Count() > 3000 && sortedGraphs.Count >= 20)
                             {
                                 // Process graphs
-                                await ProcessGraphsAsync(packageId, sortedGraphs);
+                                await ProcessGraphsAsync(packageId, sortedGraphs, _customDomain);
```
Comment (scottbommarito) [inline src/Ng/Jobs/LightningJob.cs]: You shouldn't need to pass `_customDomain` around, given that it's a field on the object.

File: src/Catalog/CatalogItem.cs (hunk 4 above)
Comment (scottbommarito) [inline src/Catalog/CatalogItem.cs]: Given that you are using this method constantly outside of the context of the catalog, I would move this to another shared static class.

File: src/Catalog/CatalogItem.cs (hunk 1 above)
Comment (cristinamanum) [inline src/Catalog/CatalogItem.cs]: Done

File: src/Catalog/CatalogWriterBase.cs (hunk 2 above)
Comment (cristinamanum) [inline src/Catalog/CatalogWriterBase.cs]: Done

File: src/Catalog/CatalogItem.cs (hunk 3 above)
Comment (cristinamanum) [inline src/Catalog/CatalogItem.cs]: Done

File: src/Catalog/CatalogItem.cs (hunk 4 above)
Comment (cristinamanum) [inline src/Catalog/CatalogItem.cs]: Done

File: src/Catalog/CatalogItem.cs (hunk 5 above)
Comment (cristinamanum) [inline src/Catalog/CatalogItem.cs]: It will be a small perf impact. If changed in the suggested way then the set of the base address needs to happen after the set of custom domain. In the current way the get will use the data at the current time.

File: src/Catalog/Registration/RegistrationMaker.cs (hunk 6 above)
Comment (cristinamanum) [inline src/Catalog/Registration/RegistrationMaker.cs]: Done

File: src/Catalog/Registration/RegistrationMakerCatalogWriter.cs (hunk 7 above)
Comment (cristinamanum) [inline src/Catalog/Registration/RegistrationMakerCatalogWriter.cs]: Done

File: src/Catalog/Registration/RegistrationPersistence.cs (hunk 8 above)
Comment (cristinamanum) [inline src/Catalog/Registration/RegistrationPersistence.cs]: Not sure what space i need to remove?

File: src/Catalog/CatalogWriterBase.cs (hunk 9 above)
Comment (cristinamanum) [inline src/Catalog/CatalogWriterBase.cs]: The feed2catalog will not be impacted by these changes as long as AppendOnlyCatalogWriter (used by feed2catalog) will not subscribe to use a customDomain.
By default customDomain is null and the ReplaceHostForUri is noop for this case. I will go some feed2catalog more testing.

File: src/Catalog/CatalogItem.cs (hunk 4 above)
Comment (cristinamanum) [inline src/Catalog/CatalogItem.cs]: Good point. Added.

File: src/Catalog/CatalogWriterBase.cs (hunk 12 above)
Comment (cristinamanum) [inline src/Catalog/CatalogWriterBase.cs]: We discussed offline and I agree that the changes could be risky.  I tried to be as less intrusive as possible  using the current design. The mental model that I followed was the separation of concerns between the storage and content. The host links in content (generated by the writers) can be different by the storage host. We can do more radical design changes but  that it will add to the time and risk.

File: src/Catalog/Registration/FlatContainerPackagePathProvider.cs (hunk 10 above)
Comment (cristinamanum) [inline src/Catalog/Registration/FlatContainerPackagePathProvider.cs]: Good suggestion to not have the v3-flatcontainer hard-coded. The code was updated to use a parameter for it.

File: src/Ng/Jobs/Catalog2RegistrationJob.cs (hunk 11 above)
Comment (cristinamanum) [inline src/Ng/Jobs/Catalog2RegistrationJob.cs]: The cursor is calculated now, but the flatcontainer uri is still passed as argument. I preferred this way instead of passing the index.json uri , open and read the flat container from there.

File: src/Ng/Jobs/LightningJob.cs (hunk 13 above)
Comment (cristinamanum) [inline src/Ng/Jobs/LightningJob.cs]: Right... updated.

File: src/Catalog/CatalogItem.cs (hunk 4 above)
Comment (cristinamanum) [inline src/Catalog/CatalogItem.cs]: It is used in the catalog content generation. IMO is to have it here or catalogwriter base. I added in the writerbase. If any strong preference let me know.

File: src/Catalog/Registration/RegistrationPersistence.cs (hunk 8 above)
Comment (scottbommarito) [inline src/Catalog/Registration/RegistrationPersistence.cs]: after the variable name
`_customDomain ;` -> `_customDomain;`

File: src/Catalog/Registration/FlatContainerPackagePathProvider.cs (hunk 10 above)
Comment (scottbommarito) [inline src/Catalog/Registration/FlatContainerPackagePathProvider.cs]: We came to the conclusion that it would be best not to use the V3 `index.json` because it's now important to have the registration blobs use the same flat-container, and using the `index.json` could lead to a mismatch.

File: src/Catalog/CatalogWriterBase.cs (hunk 9 above)
Comment (cristinamanum) [inline src/Catalog/CatalogWriterBase.cs]: Manual tested the feed2catalog job.

File: src/Catalog/Registration/RegistrationPersistence.cs (hunk 8 above)
Comment (cristinamanum) [inline src/Catalog/Registration/RegistrationPersistence.cs]: Done

File: src/Ng/lightning-template.txt [hunk 14]
```diff
@@ -6,7 +6,7 @@ title [cursorFile]
...
+start /w ng.exe lightning -command strike -indexFile "[indexFile]" -cursorFile "[cursorFile]" -contentBaseAddress "[contentBaseAddress]" -verbose "[verbose]" -storageBaseAddress "[storageBaseAddress]" -storageAccountName "[storageAccountName]" -storageKeyValue "[storageKeyValue]" -storageContainer "[storageContainer]" -useCompressedStorage "[useCompressedStorage]" -compressedStorageBaseAddress "[compressedStorageBaseAddress]" -compressedStorageAccountName "[compressedStorageAccountName]" -compressedStorageKeyValue "[compressedStorageKeyValue]" -compressedStorageContainer "[compressedStorageContainer]" -useSemVer2Storage "[useSemVer2Storage]" -semVer2StorageBaseAddress "[semVer2StorageBaseAddress]" -semVer2StorageAccountName "[semVer2StorageAccountName]" -semVer2StorageKeyValue "[semVer2StorageKeyValue]" -semVer2StorageContainer "[semVer2StorageContainer]" [optionalArguments]
```
Comment (joelverhagen) [inline src/Ng/lightning-template.txt]: We should format this such that each kvp is on its own line

Comment (joelverhagen) [inline src/Ng/lightning-template.txt]: Like the file above 😄

File: src/Catalog/Registration/RegistrationPersistence.cs [hunk 15]
```diff
@@ -200,24 +253,53 @@ static void AddExistingItem(IDictionary<RegistrationEntryKey, RegistrationCatalo
...
             Trace.TraceInformation("RegistrationPersistence.SaveLargeRegistration: registrationBaseAddress = {0} items: {1}", registrationBaseAddress, items.Count);
 
             IList<Uri> cleanUpList = new List<Uri>();
 
-            await SaveRegistration(storage, registrationBaseAddress, items, cleanUpList, null, partitionSize, contentBaseAddress, cancellationToken);
+            await SaveRegistration(storage,
+                                   registrationBaseAddress,
+                                   items,
+                                   cleanUpList,
+                                   null,
```
Comment (joelverhagen) [inline src/Catalog/Registration/RegistrationPersistence.cs]: name nulls

File: src/Catalog/CatalogWriterBase.cs [hunk 16]
```diff
@@ -217,7 +241,7 @@ protected virtual StorageContent CreateIndexContent(IGraph graph, Uri type)
 
             if (parent != null)
             {
-                graph.Assert(resourceNode, graph.CreateUriNode(Schema.Predicates.CatalogParent), graph.CreateUriNode(parent));
+                graph.Assert(resourceNode, graph.CreateUriNode(Schema.Predicates.CatalogParent), graph.CreateUriNode(ReplaceHostForUri(parent,_customDomain)));
```
Comment (joelverhagen) [inline src/Catalog/CatalogWriterBase.cs]: nit space after comma

File: src/Ng/Jobs/LightningJob.cs [hunk 17]
```diff
@@ -126,6 +133,7 @@ public override string GetUsage()
         private RegistrationStorageFactories _storageFactories;
         private ShouldIncludeRegistrationPackage _shouldIncludeSemVer2;
         private IDictionary<string, string> _arguments;
+        private string _customDomain;
```
Comment (joelverhagen) [inline src/Ng/Jobs/LightningJob.cs]: Unused?

File: tests/NgTests/RegistrationCollectorTests.cs [hunk 18]
```diff
@@ -12,6 +12,9 @@
 using NgTests.Infrastructure;
 using NuGet.Services.Metadata.Catalog;
 using NuGet.Services.Metadata.Catalog.Registration;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
```
Comment (joelverhagen) [inline tests/NgTests/RegistrationCollectorTests.cs]: Are these changes necessary?

File: src/Ng/Jobs/LightningJob.cs [hunk 19]
```diff
@@ -147,19 +155,27 @@ protected override void Init(IDictionary<string, string> arguments, Cancellation
...
             arguments[Arguments.StorageType] = Arguments.AzureStorageType;
 
-            // Enable greater HTTP parallization.
+            // Enable greater HTTP parallelization.
             ServicePointManager.DefaultConnectionLimit = CatalogIndexReader.MaxDegreeOfParallelism;
 
             // Configure the package path provider.
-            RegistrationMakerCatalogItem.PackagePathProvider = new PackagesFolderPackagePathProvider();
+            var useFlatContainerAsPackageContent = arguments.GetOrDefault<bool>(Arguments.ContentIsFlatContainer, false);
+            if (!useFlatContainerAsPackageContent)
```
Comment (chenriksson) [inline src/Ng/Jobs/LightningJob.cs]: nit, would reverse the if/else so you don't need the '!'

File: src/Ng/Jobs/LightningJob.cs [hunk 20]
```diff
@@ -73,6 +73,13 @@ public override string GetUsage()
             sw.WriteLine($"      Compressed only: Azure Storage account name.");
             sw.WriteLine($"  -{Arguments.CompressedStorageContainer} <container>");
             sw.WriteLine($"      Compressed only: Container to generate registrations in.");
+            sw.WriteLine($"  -{Arguments.ContentIsFlatContainer} true|false");
+            sw.WriteLine($"      Boolean to indicate if the registration blobs will have the content from the packages or flat container.");
+            sw.WriteLine($"  -{Arguments.FlatContainerName} <the flat container name>");
+            sw.WriteLine($"      It is required when the ContentIsFlatContainer flag is true.");
```
Comment (chenriksson) [inline src/Ng/Jobs/LightningJob.cs]: Assuming ContentIsFlatContainer is only necessary so we can have hard-coded default for FlatContainerName?

File: src/Ng/lightning-template.txt (hunk 14 above)
Comment (cristinamanum) [inline src/Ng/lightning-template.txt]: Done

File: src/Ng/Jobs/LightningJob.cs (hunk 17 above)
Comment (cristinamanum) [inline src/Ng/Jobs/LightningJob.cs]: Updated

File: tests/NgTests/RegistrationCollectorTests.cs (hunk 18 above)
Comment (cristinamanum) [inline tests/NgTests/RegistrationCollectorTests.cs]: Not needed. Updated.

File: src/Catalog/Registration/FlatContainerPackagePathProvider.cs [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            _container = container;
+        }
+        public string GetPackagePath(string id, string version)
+        {
+            version = NuGetVersionUtility.NormalizeVersion(version);
+
+            return $"{_container}/{id.ToLowerInvariant()}/{version.ToLowerInvariant()}/{id.ToLowerInvariant()}.{version.ToLowerInvariant()}.nupkg";
+        }
+
```
Comment (cristinamanum) [inline src/Catalog/Registration/FlatContainerPackagePathProvider.cs]: Done

File: src/Ng/Jobs/LightningJob.cs (hunk 20 above)
Comment (cristinamanum) [inline src/Ng/Jobs/LightningJob.cs]: It is passed as argument in order to be able to change the flatContainer container name.
